azext_iot/iothub/providers/message_endpoint.py

```diff
@@ -0,0 +1,351 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from typing import Optional
+from knack.log import get_logger
+from azure.cli.core.azclierror import (
+    ArgumentUsageError,
+    RequiredArgumentMissingError,
+    ResourceNotFoundError,
+    InvalidArgumentValueError
+)
+from azext_iot.common.embedded_cli import EmbeddedCLI
+from azext_iot.iothub.common import (
+    INVALID_CLI_CORE_FOR_COSMOS, SYSTEM_ASSIGNED_IDENTITY, AuthenticationType, EncodingFormat, EndpointType
+)
+from azext_iot.iothub.providers.base import IoTHubProvider
+from azext_iot.common._azure import parse_cosmos_db_connection_string
+from azure.mgmt.iothub.models import ManagedIdentity
+
+
+logger = get_logger(__name__)
+
+
+class MessageEndpoint(IoTHubProvider):
+    def __init__(
+        self,
+        cmd,
+        hub_name: str,
+        rg: Optional[str] = None,
+    ):
+        super(MessageEndpoint, self).__init__(cmd, hub_name, rg, dataplane=False)
+        self.support_cosmos = hasattr(self.hub_resource.properties.routing.endpoints, "cosmos_db_sql_collections")
+        self.cli = EmbeddedCLI(cli_ctx=self.cmd.cli_ctx)
+
+    def create(
+        self,
+        endpoint_name: str,
+        endpoint_type: str,
+        endpoint_account_name: Optional[str] = None,
+        endpoint_resource_group: Optional[str] = None,
+        endpoint_subscription_id: Optional[str] = None,
+        endpoint_policy_name: Optional[str] = None,
+        connection_string: Optional[str] = None,
+        container_name: Optional[str] = None,
+        encoding: Optional[str] = None,
+        batch_frequency: int = 300,
+        chunk_size_window: int = 300,
+        file_name_format: str = '{iothub}/{partition}/{YYYY}/{MM}/{DD}/{HH}/{mm}',
+        endpoint_uri: Optional[str] = None,
+        entity_path: Optional[str] = None,
+        database_name: Optional[str] = None,
+        primary_key: Optional[str] = None,
+        secondary_key: Optional[str] = None,
+        partition_key_name: Optional[str] = None,
+        partition_key_template: Optional[str] = None,
+        identity: Optional[str] = None
+    ):
+        resource_group_name = self.hub_resource.additional_properties["resourcegroup"]
+        if not endpoint_resource_group:
+            endpoint_resource_group = resource_group_name
```

Comment:
tiny nit - i don't think `resource_group_name` is used outside this scope:
```suggestion
        if not endpoint_resource_group:
            endpoint_resource_group = self.hub_resource.additional_properties["resourcegroup"]
```

Reply:
it is also used in the begin_create_or_update but I can just replace the variable